### PR TITLE
feat(units): persist cloud container addresses and ports

### PR DESCRIPTION
**Pull request by wallyworld**
Adds state code to the application domain to persist cloud container address and ports. 

To save a cloud container address, a placeholder link layer device is created to tie the address to the container. Because it's a placeholder, it doesn't have a MAC address so this column is not optional.

Also includes code to support managing the various related enums - address scope, origin etc, device type etc.

Drive by fix to use a string rather than a string pointer for unit names. The unit name will end up being supplied by the service layer, either created from the k8s pod number or by using a sequence number for vm units.

## QA steps

application domain unit tests

## Links

**Jira card:** [JUJU-6586](https://warthogs.atlassian.net/browse/JUJU-6586)

### CODE REVIEW

File: domain/application/state/application.go [hunk 1]
```diff
@@ -683,18 +827,72 @@ DELETE FROM net_node WHERE uuid IN
...
+	}
+	deleteAddressStmt, err := st.Prepare(`
+DELETE FROM ip_address
+WHERE device_uuid IN
+(SELECT device_uuid FROM link_layer_device lld
+WHERE lld.net_node_uuid = $coreUnit.net_node_uuid)
+`, unit)
+	if err != nil {
 		return errors.Trace(err)
 	}
```
Comment (hpidcock) [inline domain/application/state/application.go:872]: Sometimes it is hard to grok these state methods because the prepares are all done first then the statement execution later. In this case the ordering of the prepares made me double take because it doesn't match the order of the statement execution. That is on me, nothing is actually wrong here (but in my head the CTE here was trying to select a device_uuid for a link_layer_device that was already deleted :sweat_smile: )

Comment (hpidcock) [reply domain/application/state/application.go:872]: This is a me problem, not a you problem, I'm not used to reviewing state stuff yet because I haven't written any myself.

File: domain/application/state/state.go [hunk 2]
```diff
@@ -737,6 +737,11 @@ WHERE uuid = $charmID.uuid;
 		return nil, -1, fmt.Errorf("failed to get charm lxd profile: %w", err)
 	}
 
+	// Sometimes we get am empty slice, sometimes a nil slice.
```
Comment (hpidcock) [inline domain/application/state/state.go:740]:
```suggestion
	// Sometimes we get an empty slice, sometimes a nil slice.
```

File: domain/application/state/state.go [hunk 3]
```diff
@@ -737,6 +737,11 @@ WHERE uuid = $charmID.uuid;
...
+	// Cater for both cases.
+	if len(profile.LXDProfile) == 0 {
```
Comment (hpidcock) [inline domain/application/state/state.go:742]: This is annoying.

File: domain/ipaddress/ipaddress.go [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+	case corenetwork.ConfigLoopback:
+		return ConfigTypeLoopback
+	}
+	return ConfigTypeUnknown
+}
+
+// Scope represents the scope of
+// an IP address, as recorded in
+// the ip_address_scope lookup table.
+type Scope int
```
Comment (hpidcock) [inline domain/ipaddress/ipaddress.go:41]: Ahh you must be using your VT05 in split screen.
```suggestion
// Scope represents the scope of an IP address, as recorded in the ip_address_scope lookup table.
type Scope int
```

or even
```suggestion
// Scope represents the scope of an IP address, as recorded in the
// ip_address_scope lookup table.
type Scope int
```

File: domain/ipaddress/ipaddress.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package ipaddress
```
Comment (hpidcock) [inline domain/ipaddress/ipaddress.go:4]: Maybe add a doc.go?

File: domain/linklayerdevice/device.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package linklayerdevice
```
Comment (hpidcock) [inline domain/linklayerdevice/device.go:4]: Maybe add a doc.go?

**Review by hpidcock — APPROVED** [review]
Approved with some nitpicks. Needs a second review, maybe @tlm.

File: domain/ipaddress/ipaddress.go (hunk 4 above)
Comment (wallyworld) [reply domain/ipaddress/ipaddress.go]: ok

File: domain/linklayerdevice/device.go (hunk 6 above)
Comment (wallyworld) [reply domain/linklayerdevice/device.go:4]: ok

File: domain/ipaddress/ipaddress.go (hunk 5 above)
Comment (wallyworld) [reply domain/ipaddress/ipaddress.go:4]: ok

File: domain/application/state/state.go (hunk 3 above)
Comment (wallyworld) [reply domain/application/state/state.go:743]: Yeah, I added a TODO to figure out why

File: domain/application/state/application.go (hunk 1 above)
Comment (wallyworld) [reply domain/application/state/application.go]: Fair point. I'll make the order of prepare match the order of execution.

File: domain/application/state/application.go [hunk 7]
```diff
@@ -575,8 +558,169 @@ ON CONFLICT(net_node_uuid) DO UPDATE
...
+		return errors.Trace(err)
+	}
+
+	// Container addresses are never deleted unless the container itself is deleted.
+	// First see if there's an existing address recorded.
+	if err := tx.Query(ctx, selectAddressUUIDStmt, ipAddr).Get(&ipAddr); err != nil {
+		if !errors.Is(err, sqlair.ErrNoRows) {
+			return fmt.Errorf("querying existing cloud container address for device %q: %w", deviceID, err)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/state/application.go]: ok

File: domain/application/state/application.go [hunk 8]
```diff
@@ -575,8 +558,169 @@ ON CONFLICT(net_node_uuid) DO UPDATE
...
+	if errors.Is(err, sqlair.ErrNoRows) {
+		deviceUUID, err := uuid.NewUUID()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		cloudContainerDeviceInfo.UUID = deviceUUID.String()
+		if err := tx.Query(ctx, insertCloudContainerDeviceStmt, cloudContainerDeviceInfo).Run(); err != nil {
+			return errors.Annotatef(err, "inserting cloud container device for unit %q", unitName)
+		}
+		deviceID = cloudContainerDeviceInfo.UUID
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/state/application.go]: ok

File: domain/application/service/application.go [hunk 9]
```diff
@@ -311,18 +316,28 @@ func makeUpsertUnitArgs(in AddUnitArg) application.UpsertUnitArg {
 			// TODO(units) - handle the in.CloudContainer.Address space ID
 			// For k8s we'll initially create a /32 subnet off the container address
 			// and add that to the default space.
-			result.CloudContainer.Address = &application.Address{
+			result.CloudContainer.Address = &application.ContainerAddress{
+				// For cloud containers, the device is a placeholder without
+				// a MAC address and once inserted, not updated. It just exits
+				// to tie the address to the net node corresponding to the
+				// cloud container.
+				Device: application.ContainerDevice{
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/service/application.go:324]: It will be implicitly used when addresses are retrieved. There's a test which validates addresses are created as expected using this SQL
```
SELECT address_value, type_id, scope_id, origin_id FROM ip_address ipa
JOIN link_layer_device lld ON lld.uuid = ipa.device_uuid
JOIN unit u ON u.net_node_uuid = lld.net_node_uuid WHERE u.name = ?
```

File: domain/application/state/application.go [hunk 10]
```diff
@@ -575,8 +558,166 @@ ON CONFLICT(net_node_uuid) DO UPDATE
...
+
+	// Container addresses are never deleted unless the container itself is deleted.
+	// First see if there's an existing address recorded.
+	err = tx.Query(ctx, selectAddressUUIDStmt, ipAddr).Get(&ipAddr)
+	if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return fmt.Errorf("querying existing cloud container address for device %q: %w", deviceID, err)
+	}
+
+	// Create a UUID for new addresses.
+	newAddr := ipAddr.AddressUUID == ""
```
Comment (tlm) [inline domain/application/state/application.go:668]: Poorly named. `newAddr` sounds like this is a new address that you have made not a flag to tell us if we should make a new address uuid.

In any event uuid generation should be pushed upwards like the previous comments.

File: domain/application/state/application.go [hunk 11]
```diff
@@ -683,21 +824,75 @@ DELETE FROM net_node WHERE uuid IN
...
 }
 
+func (st *ApplicationState) deleteCloudContainerAddresses(ctx context.Context, tx *sqlair.TX, netNodeID string) error {
+	unit := coreUnit{
+		NetNodeID: netNodeID,
+	}
+	deleteAddressStmt, err := st.Prepare(`
+DELETE FROM ip_address
+WHERE device_uuid IN
+(SELECT device_uuid FROM link_layer_device lld
```
Comment (tlm) [inline domain/application/state/application.go:858]: The form I have seen us following for this is to leave the select on the same line as the in and indent everything up to the select.

**Review by tlm — COMMENTED** [review]
Thanks @wallyworld,

Initial grok, think I need to refresh my head and go for another pass now I have it all down.

I am concerned that the unit name in the ddl should just be the number and not have the application. But that is a side bar conversation and not related to this PR.

File: apiserver/facades/client/application/deploy.go [hunk 12]
```diff
@@ -148,8 +148,7 @@ func DeployApplication(
 	}
 	unitArgs := make([]applicationservice.AddUnitArg, args.NumUnits)
 	for i := 0; i < args.NumUnits; i++ {
-		n := fmt.Sprintf("%s/%d", args.ApplicationName, nextUnitNum+i)
-		unitArgs[i].UnitName = &n
+		unitArgs[i].UnitName = fmt.Sprintf("%s/%d", args.ApplicationName, nextUnitNum+i)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply apiserver/facades/client/application/deploy.go:151]: Yes, there's been a few conversations already. One early line of though was that we'd increment a unit number column but that doesn't guarantee a monotonically increasing number. The same applies to machine etc. We'll need to implement a namespaced sequence.

File: domain/application/service/application.go [hunk 13]
```diff
@@ -311,18 +316,28 @@ func makeUpsertUnitArgs(in AddUnitArg) application.UpsertUnitArg {
...
 			// and add that to the default space.
-			result.CloudContainer.Address = &application.Address{
+			result.CloudContainer.Address = &application.ContainerAddress{
+				// For cloud containers, the device is a placeholder without
+				// a MAC address and once inserted, not updated. It just exists
+				// to tie the address to the net node corresponding to the
+				// cloud container.
+				Device: application.ContainerDevice{
+					Name:              fmt.Sprintf("placeholder for %q cloud container", in.UnitName),
+					DeviceTypeID:      linklayerdevice.DeviceTypeUnknown,
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/service/application.go:326]: These are a placeholder and previous conversations agreed on this approach. As far as I am aware don't have visibility to the underlying link layer device like we do for vms. We want a consistent data model so need to fill in the blanks so to speak.

File: domain/application/service/application.go [hunk 14]
```diff
@@ -561,7 +578,7 @@ func (s *ApplicationService) insertCAASUnit(
 	}
 	if orderedID >= appScale.Scale ||
 		(appScale.Scaling && orderedID >= appScale.ScaleTarget) {
-		return fmt.Errorf("unrequired unit %s is not assigned%w", *args.UnitName, errors.Hide(applicationerrors.UnitNotAssigned))
+		return fmt.Errorf("unrequired unit %s is not assigned%w", args.UnitName, errors.Hide(applicationerrors.UnitNotAssigned))
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/service/application.go:581]: I don't want to piecemeal it. Greenfields code yes, but having two imports for "errors" will suck. We need to separately sweep across existing code and do the changes a package at a time.

File: domain/application/state/application.go [hunk 15]
```diff
@@ -400,16 +399,16 @@ func (st *ApplicationState) InsertUnit(
 	ctx domain.AtomicContext, appID coreapplication.ID, args application.UpsertUnitArg,
 ) error {
 	// Should not happen, defensive check.
-	if args.UnitName == nil {
+	if args.UnitName == "" {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/state/application.go:402]: This was a carry over and was supposed to be deleted since there's a not null constraint but obviously slipped through the net. Will fix.

File: domain/application/state/application.go [hunk 16]
```diff
@@ -450,23 +449,17 @@ func (st *ApplicationState) insertUnit(
...
+		return errors.Annotatef(err, "creating net node row for unit %q", args.UnitName)
 	}
 	if err := tx.Query(ctx, createUnitStmt, createParams).Run(); err != nil {
-		return errors.Annotatef(err, "creating unit row for unit %q", unitName)
+		return errors.Annotatef(err, "creating unit row for unit %q", args.UnitName)
 	}
 	if args.CloudContainer != nil {
-		if err := st.upsertUnitCloudContainer(ctx, tx, unitName, nodeUUID.String(), args.CloudContainer); err != nil {
-			return errors.Annotatef(err, "creating cloud container row for unit %q", unitName)
+		if err := st.upsertUnitCloudContainer(ctx, tx, args.UnitName, nodeUUID.String(), args.CloudContainer); err != nil {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/state/application.go:461]: This has been a topic of conversation - we had agreed that it's fine for optional attributes to be conditionally persisted in the state layer, since there's no transformation occurring.

File: domain/application/state/application.go (hunk 10 above)
Comment (wallyworld) [reply domain/application/state/application.go]: Fixed name

File: domain/application/state/application.go [hunk 17]
```diff
@@ -575,8 +558,166 @@ ON CONFLICT(net_node_uuid) DO UPDATE
...
+INSERT INTO link_layer_device (*) VALUES ($cloudContainerDevice.*)
+`, cloudContainerDeviceInfo)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	// See if the link layer device exists, if not insert it.
+	var deviceID string
+	err = tx.Query(ctx, selectCloudContainerDeviceStmt, cloudContainerDeviceInfo).Get(&cloudContainerDeviceInfo)
+	if errors.Is(err, sqlair.ErrNoRows) {
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/state/application.go:612]: inverted, still some indentation though

File: domain/application/service/application.go [hunk 18]
```diff
@@ -287,7 +289,10 @@ func (s *ApplicationService) CreateApplication(
 
 	unitArgs := make([]application.UpsertUnitArg, len(units))
 	for i, u := range units {
-		unitArgs[i] = makeUpsertUnitArgs(u)
+		unitArgs[i], err = makeUpsertUnitArgs(u)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/service/application.go:292]: IIANM we've agreed upsert is fine in the state layer, but should not be exposed outside the service layer.

File: domain/application/state/application.go [hunk 19]
```diff
@@ -450,23 +445,17 @@ func (st *ApplicationState) insertUnit(
...
+		return errors.Annotatef(err, "creating net node row for unit %q", args.UnitName)
 	}
 	if err := tx.Query(ctx, createUnitStmt, createParams).Run(); err != nil {
-		return errors.Annotatef(err, "creating unit row for unit %q", unitName)
+		return errors.Annotatef(err, "creating unit row for unit %q", args.UnitName)
 	}
 	if args.CloudContainer != nil {
-		if err := st.upsertUnitCloudContainer(ctx, tx, unitName, nodeUUID.String(), args.CloudContainer); err != nil {
-			return errors.Annotatef(err, "creating cloud container row for unit %q", unitName)
+		if err := st.upsertUnitCloudContainer(ctx, tx, args.UnitName, nodeUUID.String(), args.CloudContainer); err != nil {
```
Comment (tlm) [inline domain/application/state/application.go:457]: Small annoyance here. I think nodeUUID should be passed to `upsertUnitCloudContainer` as the uuid type and not a string. By the caller changing the type of the value it is starting to make decisions on behalf of the function that it is delegating decision making to.

TLDR keep the strong type all the way through until it reaches the thing that is actually responsible for translation.

**Review by tlm — COMMENTED** [review]
Just a few comments on this round. Happy to give it approval after this.

File: domain/application/service/application.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (tlm) [reply domain/application/service/application.go:292]: Talked and agreed this is fine.

**Review by tlm — APPROVED** [review]
lgtm

File: domain/application/state/application.go (hunk 19 above)
Comment (wallyworld) [reply domain/application/state/application.go:457]: yeah, was aiming for consistency. we already have to change ID back to UUID so was expecting this would be done as part of that separate refactoring

File: domain/application/service/application.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/service/application.go:326]: Discussed f2f, this aligns with the existing Juju model (see link layer device types in core). We could change but need to do it holistically.

File: domain/application/service/application.go [hunk 20]
```diff
@@ -532,10 +547,12 @@ func (s *ApplicationService) RegisterCAASUnit(ctx context.Context, appName strin
...
 		p.CloudContainer.AddressOrigin = &origin
 	}
-	// We need to do a bunch of business logic in the one transaction so pass in a closure that is
-	// given the transaction to use.
-	unitArg := makeUpsertUnitArgs(p)
-	err := s.st.RunAtomic(ctx, func(ctx domain.AtomicContext) error {
+
+	unitArg, err := makeUpsertUnitArgs(p)
+	if err != nil {
+		return errors.Annotatef(err, "composing unit %q: %v", args.UnitName, err)
```
[the remark this answers is not in the document]
Comment (wallyworld) [reply domain/application/service/application.go:553]: Hmmm, I'm not sure the type of unit matter. It's just a unit of an application.  But I've fixed the error messages to be aligned.

File: domain/application/state/application.go [hunk 21]
```diff
@@ -575,8 +554,163 @@ ON CONFLICT(net_node_uuid) DO UPDATE
...
+	}
+
+	// Update the address values.
+	if err = tx.Query(ctx, upsertAddressStmt, ipAddr).Run(); err != nil {
+		return fmt.Errorf("updating cloud container address attributes for device %q: %w", deviceUUID, err)
+	}
+	return nil
+}
+
+type ports []string
```
Comment (SimonRichardson) [inline domain/application/state/application.go:677]: This should be in the types.go file.

File: domain/application/state/state.go (hunk 3 above)
Comment (SimonRichardson) [reply domain/application/state/state.go:743]: This is because sqlite and dqlite differ. You can only run state tests with dqlite lite because they handle empty byte slices different. 

Remove this if statement.

File: domain/application/state/application.go [hunk 22]
```diff
@@ -683,29 +818,85 @@ DELETE FROM net_node WHERE uuid IN
...
+	}
+	deleteAddressStmt, err := st.Prepare(`
+DELETE FROM ip_address
+WHERE device_uuid IN (
+    SELECT device_uuid FROM link_layer_device lld
+    WHERE lld.net_node_uuid = $coreUnit.net_node_uuid
+)
+`, unit)
+	if err != nil {
+		return errors.Trace(err)
```
Comment (manadart) [inline domain/application/state/application.go:857]: I thought we weren't going to do this any more.
